BatchedMesh: Add setCustomSort function, hybrid radix sort implementation to examples

gkjohnson opened the pull request:
Related issue: #22376, https://github.com/mrdoob/three.js/pull/27202#issuecomment-1817640271

**Description**

For the sake of discussion here's a draft PR that shows the performance improvements using the suggested sort implementation from https://github.com/mrdoob/three.js/pull/27202#issuecomment-1817640271. If you want to try it you can run the batched example page and toggle the "useRadix" option to see the difference. The sort time is logged in the console.

**Performance**

Overall it looks like the new sort is providing roughly a 10x improvement in sort time over the native sort information. The drawback of the sort is that it requires more memory to store a temporary array:

| | hybrid sort | Array.sort |
|---|---|---|
| sort time | ~0.65ms | ~6.3ms |
| fps | ~94fps | ~74fps |

There seem to be occasional dips in fps with the hybrid sort to ~89fps on my machine but I'm not sure why.

Overall it's a meaningful improvement in complex cases. I'm not sure what opinions of others are on whether this should be included in three, though.

cc @Mugen87 @mrdoob @sciecode

gkjohnson committed:
Add radix sort to batched mesh

CodyJasonBennett commented:
> There seem to be occasional dips in fps with the hybrid sort to ~89fps on my machine but I'm not sure why.

I haven't delved much into the implementation yet, but my first thought is that maybe garbage collection is playing a role here.

gkjohnson committed:
Fix options generation

gkjohnson commented:
> I haven't delved much into the implementation yet, but my first thought is that maybe garbage collection is playing a role here.

That was my first thought, too - I had missed that an array was getting created internally on every call when I thought was cached so that was likely it. Just pushed the fix.

sciecode commented:
I get similar sorting times, but slightly better overall performance ( ~40% )

|   | hybrid sort | Array.sort |
| - | ----------- | ---------- |
| sort | ~0.69ms | ~6.5ms |
| fps | ~89fps | ~62fps |

I believe radix-sort can be a beneficial addition to three, but it's important to inform the users of the restrictions, as it is a less general function than `Array.sort`.

There are currently two important restrictions:
- It expects values in range [0, UINT32_MAX]
- `array.length` <= UINT32_MAX

The first restriction is not bad as it might seem, it does work on floating-point numbers if they are normalized to expected range, as Garrett did in the example. And the same thing can be done with signed 32-bit integers, like so:
```
// UINT32_MAX = 4_294_967_295
// INT32_MAX+1 = 2_147_483_648
const remap = (signed) => signed + INT32_MAX + 1;
```

sciecode commented on src/objects/BatchedMesh.js:
```diff
@@ -967,7 +968,10 @@ class BatchedMesh extends Mesh {
 
 						// get the distance from camera used for sorting
-						const z = _vector.distanceToSquared( _sphere.center );
+						// radix sort operates on 32 bit signed integers
```
unsigned https://github.com/mrdoob/three.js/pull/27213#issuecomment-1817905549

Mugen87 commented:
TBH, I would prefer to keep this outside of the core since `BatchedMesh` is already a quite complex component and we should not over-engineer it. 

Maybe you could add something like `BatchedMesh.setCustomSort()` so every developer cann add a custom sort implementation if performance would be an issue.

gkjohnson commented:
> TBH, I would prefer to keep this outside of the core since BatchedMesh is already a quite complex component and we should not over-engineer it.
>
> Could you add something like BatchedMesh.setCustomSort() so every developer can add a custom sort implementation if performance would be an issue?

It's a pretty significant improvement but admittedly likely only relevant for particularly complex cases. How would you feel about something like a `fastBatchedMeshSort` and / or `hybridRadixSort` to the three.js examples so this code is available?

Mugen87 commented:
So you mean moving `src/objects/radixSort.js` to the addons and make a new example that shows how to use it via `setCustomSort()`? That sounds good to me 👍 .

gkjohnson committed:
Add custom sort function, separate radix sort function to examples

gkjohnson committed:
Adjust custom sort application

gkjohnson committed:
Merge remote-tracking branch 'mrdoob/dev' into radix-test

gkjohnson committed:
revert count and position in demo

gkjohnson committed:
Update demo

gkjohnson committed:
Add docs

gkjohnson commented:
I've adjusted the PR to move the radix sort function into a "SortUtils.js" example file, added a "customSort" function to BatchedMesh, and updated the example to include a "useCustomSort" function.

gkjohnson committed:
Comments
